sal/tests/server/unit/test_auth.py

```diff
@@ -0,0 +1,257 @@
+                                          mock_authentication_provider,
+                                          auth_required):
+
+    """
+    GIVEN
+        An authentication provider is defined
+    WHEN
+        Whether authentication is required is requested  
+    THEN
+        Whether authentication is required is returned
+    """
+
+    mock_authentication_provider.required.return_value = auth_required
+    assert auth.auth_required() == auth_required
+
+
+@settings(suppress_health_check=[HealthCheck.function_scoped_fixture])
+@given(addr=st.ip_addresses(),
+       user=st.text(alphabet=st.characters(whitelist_categories=('Lu',
+                                                                 'Ll',
+                                                                 'Nd',
+                                                                 'P'))))
+def test_verify_valid_token(addr, user, server_with_auth_prov):
+
+    """
+    Uses `generate_token` to create token, as the essential behaviour to test
+    is that a generated token can correctly verified 
+
+    GIVEN
+        A server with a secret token key
+        AND a server with a token lifetime
+        AND a request from a IP address
+        AND a token generated using the IP
+        AND a token generated less than the token lifetime ago
+        AND a token generated using the secret key
+        AND a token generated with a username
+    WHEN
+        The token is verified
+    THEN
+        The username is extracted from the token
+    """
+
+    # str cast required because hypothesis generates IPv4Address and
+    # IPv6Address objects
+    with server_with_auth_prov.test_request_context(
+        environ_base={'REMOTE_ADDR':str(addr)}):
+        token = auth.generate_token(user)
+        assert user == auth.verify_token(token)
+
+
+def test_verify_expired_token(server_with_auth_prov):
+
+    """
+    GIVEN
+        A server with a secret token key
+        AND a server with a token lifetime
+        AND a request from a IP address
+        AND a token generated using the IP
+        AND a token generated more than the token lifetime ago
+        AND a token generated using the secret key
+        AND a token generated with a username
+    WHEN
+        The token is verified
+    THEN
+        None is returned
+    """
+
+    user = 'username'
+    server_with_auth_prov.config['SAL']['TOKEN_LIFETIME'] = 0
+
+    with server_with_auth_prov.test_request_context(
+        environ_base={'REMOTE_ADDR':'0.0.0.0'}):
+        token = auth.generate_token(user)
+        sleep(1)
+        assert auth.verify_token(token) is None
+
+
+def test_verify_token_different_ip(server_with_auth_prov):
+
+    """
+    GIVEN
+        A server with a secret token key
+        AND a server with a token lifetime
+        AND a request from a IP address
+        AND a token generated using a different IP
+        AND a token generated more than the token lifetime ago
```

Comment:
**issue**: The token lifetime does not appear to be tested in this test. I'm guessing this is a copy and paste error from the above test.

Reply:
Well spotted!